src/overlays/actors/ovl_Boss_Tw/z_boss_tw.c

```diff
@@ -156,183 +187,5337 @@ static ColliderCylinderInit D_8094A840 = {
+                    Audio_PlayActorSound2(&this->actor, NA_SE_EN_TWINROBA_MS_FIRE - SFX_FLAG);
+                } else {
+                    Audio_PlayActorSound2(&this->actor, NA_SE_EN_TWINROBA_MS_FREEZE - SFX_FLAG);
+                }
+            }
+
+            this->flameRotation += this->updateRate1 * 0.0025f;
+            Math_ApproachF(&this->spawnPortalScale, 0.0f, 0.1f, this->updateRate1);
+            Math_ApproachF(&this->updateRate1, 50.0f, 1.0f, 2.0f);
+        }
+
+        if (Animation_OnFrame(&this->skelAnime, this->workf[ANIM_SW_TGT])) {
+            Animation_MorphToLoop(&this->skelAnime, &D_06009398, 0.0f);
+            this->workf[ANIM_SW_TGT] = 10000.0f;
+        }
+
+        if (this->timers[1] == 1) {
+            Animation_MorphToPlayOnce(&this->skelAnime, &D_06003614, 0.0f);
+            this->workf[ANIM_SW_TGT] = Animation_GetLastFrame(&D_06003614);
+            this->unk_4DC = 0.0f;
+            this->spawnPortalAlpha = 0.0f;
+            this->flameAlpha = 0.0f;
+            sBeamDivertTimer = 0;
+        }
+    } else {
+        if (Animation_OnFrame(&this->skelAnime, this->workf[ANIM_SW_TGT])) {
+            Animation_MorphToLoop(&this->skelAnime, &D_06003E34, 0.0f);
+            this->workf[ANIM_SW_TGT] = 10000.0f;
+        }
+
+        if (Animation_OnFrame(&this->skelAnime, this->workf[ANIM_SW_TGT] - 5.0f)) {
+            this->beamShootState = 0;
+            sEnvType = this->actor.params + 1;
+        }
+
+        if (Animation_OnFrame(&this->skelAnime, this->workf[ANIM_SW_TGT] - 13.0f)) {
+            Audio_PlayActorSound2(&this->actor, NA_SE_EN_TWINROBA_THROW_MASIC);
+            Audio_PlayActorSound2(&this->actor, NA_SE_EN_TWINROBA_SHOOT_VOICE);
+        }
+
+        xDiff = this->targetPos.x - this->beamOrigin.x;
+        yDiff = this->targetPos.y - this->beamOrigin.y;
+        zDiff = this->targetPos.z - this->beamOrigin.z;
+
+        this->beamYaw = Math_FAtan2F(xDiff, zDiff);
+        this->beamPitch = -Math_FAtan2F(yDiff, sqrtf(SQ(xDiff) + SQ(zDiff)));
+
+        switch (this->beamShootState) {
+            case -1:
+                break;
+            case 0:
+                if (this->timers[0] != 0) {
+                    s32 beamReflection = BossTw_CheckBeamReflection(this, globalCtx);
+                    if (beamReflection == 1) {
+                        Vec3f pos;
+                        Vec3f velocity;
+                        Vec3f accel = { 0.0f, 0.0f, 0.0f };
+
+                        for (i = 0; i < ARRAY_COUNT(sTWEffects); i++) {
+                            velocity.x = Rand_CenteredFloat(15.0f);
+                            velocity.y = Rand_CenteredFloat(15.0f);
+                            velocity.z = Rand_CenteredFloat(15.0f);
+                            pos = player->bodyPartsPos[15];
+                            BossTw_AddDotEffect(globalCtx, &pos, &velocity, &accel, (s16)Rand_ZeroFloat(2.0f) + 5,
+                                                this->actor.params, 150);
```

Comment:
this isn't the only time it happens in the file, there's a AddEffect call under a loop but that AddEffect also loops. So basically this is trying to add 150*150=22500 effects (but it's fine because the AddEffect function is never passed more effects to add than what the array can handle), `@ bug` still?

Reply:
talking with roman in my livestream, we agree that its extremely inefficient, but not a bug per se. Cause its handled properly to not go over.
Probably worth noting down in a more documentation heavy pass, incase anyone feels like trying to optimize things lol

Reply:
actaully roman noted that this effect function doesn't add 150 new effects. Its just looping over the slots to find an empty one, then it breaks. 
Still could be a source of performance issues though, cause its alot of loops

Reply:
roman is suggesting to go with `countLimit` for the arg instead of `effectCnt` to make that more clear, and I agree

Reply:
I still think it's a `@bug` because the loop running 150 times means 150 randomized effect parameters, but only the first parameters (first loop iteration) will be actually used, to fill up all of the effects buffer

I'm thinking they wanted to have a countLimit=150 of randomized effects, not 150 times the same effect

Reply:
Id personally still say thats just extra inefficient, but not a bug necessarily. Id be ok with a comment pointing this out added later, but dont know if I agree with a bug tag